## STAT 540 Seminar 03: R graphics - 'lattice' version
### Hao Chen
__2014-02-10__

* <dim id="1a">[Load the photoRec dataset and required packages](#1b)
* <dim id="2a">[The "you try" questions](#2b)
* <dim id="3a">[Take-home problem](#3b)

```{r include = FALSE}
opts_chunk$set(tidy = F, warning=F, comment=NA, prompt=T)
```

### <dim id="1b">[Load the photoRec dataset and required packages](#1a)
The following datasets will be covered in this seminar:
* [GSE4051_MINI.rds](https://github.com/jennybc/stat540_2014/blob/master/examples/photoRec/data/GSE4051_MINI.rds): mini gene expression dataset
* [GSE4051_data.tsv](https://github.com/jennybc/stat540_2014/blob/master/examples/photoRec/data/GSE4051_data.tsv): full data matrix 
* [GSE4051_design.rds](https://github.com/jennybc/stat540_2014/blob/master/examples/photoRec/data/GSE4051_design.rds): experimental design

Load required datasets:
```{r}
kDat <- readRDS("GSE4051_MINI.rds")
str(kDat)
```

```{r}
prDat <- read.table("GSE4051_data.tsv")
str(prDat, max.level = 0)
```

```{r}
prDes <- readRDS("GSE4051_design.rds")
str(prDes)
```

Load required R packages:
```{r}
library(lattice)   # for plotting
library(gridExtra)   # for grid graphics
library(RColorBrewer)  # for color palettes  
library(gplots)  # for heatmap
```

### <dim id="2b">[The "you try" questions](#2a)
[You try 1](http://ugrad.stat.ubc.ca/~stat540/seminars/seminar03_graphics-lattice.html#scatterplots): request a scatterplot of the variable poisonFang against crabHammer.
```{r}
xyplot(poisonFang ~ crabHammer, kDat)
```

[You try 2](http://ugrad.stat.ubc.ca/~stat540/seminars/seminar03_graphics-lattice.html#stripplot): Remake this plot but instead of conveying genotype via color, show developmental stage.

```{r}
nDat <- with(kDat, data.frame(sidChar, sidNum, devStage, gType, crabHammer, 
                              probeset = factor(rep(c("eggBomb", "poisonFang"), 
                                                    each = nrow(kDat))), 
                              geneExp = c(eggBomb, poisonFang)))

xyplot(geneExp ~ crabHammer | probeset, nDat, grid = TRUE, groups = devStage, auto.key = TRUE)
```

[You try 3](http://ugrad.stat.ubc.ca/~stat540/seminars/seminar03_graphics-lattice.html#stripplot): Use `densityplot()` to explore the gene expression distribution by gene and/or developmental stage. Play with 'bw' and 'n' if you like.

```{r fig.width=9, fig.height=9}
oDat <- with(kDat, data.frame(sidChar, sidNum, devStage, gType,
                              probeset = factor(rep(c("crabHammer", "eggBomb", "poisonFang"),
                                                    each = nrow(kDat))),
                              geneExp = c(crabHammer, eggBomb, poisonFang)))

denplotfun <- function(jBw, jn){
  densityplot(~ geneExp, oDat, groups = gType, auto.key = TRUE, 
              bw = jBw, n = jn, main = paste("bw =", jBw, ", n =", jn))
}

fig1 <- denplotfun(0.2, 100)
fig2 <- denplotfun(0.2, 400)
fig3 <- denplotfun(0.5, 100)
fig4 <- denplotfun(0.5, 400)

grid.arrange(fig1, fig2, fig3, fig4, nrow=2)
```

### <dim id="3b">[Take-home problem](#3a)
[Take-home problem](http://ugrad.stat.ubc.ca/~stat540/seminars/seminar03_graphics-lattice.html#take-home-problem): The full photoRec dataset has 39 samples and 29949 probesets. Choose 2 ... or 20 ... or 200 random probesets/genes and look for gene expression differences between the two genotypes, wild type versus knockout. Make use of the graphing techniques discussed this week such as scatter plots, data heatmaps, correlation heatmaps, etc. 

Draw 30 probesets at random.
```{r}
set.seed(540)
yo <- sample(1:nrow(prDat), size = 30)
hDat <- prDat[yo, ]
hDat <- as.matrix(t(hDat))
rownames(hDat) <- with(prDes, paste(devStage, gType, sidChar, sep="_"))
str(hDat)
```

### Scatter plot
Randomly draw 4 samples from 39 samples intotal
```{r fig.width=9, fig.height=9}
set.seed(540)
yo <- sample(1:nrow(hDat), size = 4)
pairDat <- t(hDat[yo, ])

panel.cor <- function(x, y, digits=2, cex.cor)
{
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y))
  txt <- format(c(r, 0.123456789), digits=digits)[1]
  text(0.5, 0.5, paste("r=",txt), cex=1.5)
}

panel.hist <- function(x, ...)
{
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(usr[1:2], 0, 1.5) )
  h <- hist(x, plot = FALSE)
  breaks <- h$breaks; nB <- length(breaks)
  y <- h$counts; y <- y/max(y)
  rect(breaks[-nB], 0, breaks[-1], y, ...)
}

pairs(pairDat, upper.panel=panel.cor, diag.panel=panel.hist)
```

### Data heatmap
```{r}
jBuPuFun <- colorRampPalette(brewer.pal(n = 9, "BuPu"))
heatmap.2(hDat, col = jBuPuFun, trace = "none")
```

### Correlation heatmap
```{r}
heatmap.2(cor(t(hDat)), col = jBuPuFun, trace = "none") 
```


